setup_unit_tests.py

```diff
@@ -0,0 +1,488 @@
+            # one of the tests wants to test if Objects are supported, but miniwdl doesn't support objects
+            # defer to the other more simpler parser
+            return extract_output_types_defer(wdl_file)
+        else:
+            raise RuntimeError("Likely unsupported type. Failed to parse WDL file %s!", wdl_file) from e
+    except Exception as e:
+        raise RuntimeError(f"Failed to parse WDL file {str(wdl_file)}!") from e
+
+    target: Union[WDL.Tree.Workflow, WDL.Tree.Task]
+
+    if document.workflow:
+        target = document.workflow
+    elif len(document.tasks) == 1:
+        target = document.tasks[0]
+    else:
+        raise Exception(
+            "Multiple tasks founds with no workflow!")  # this in theory shouldn't be hit, but the source repo has a suspicious --tasks argument passed to miniwdl sometimes
+
+    var_types = dict()
+    if target.outputs is not None:
+        for output in target.outputs:
+            var_types[output.name] = wdl_type_to_string(output.type)
+
+    return var_types
+
+
+def convert_typed_output_values(output_values: Union[None, str, Dict[str, Any], List[Any]], output_type: WDL.Type.Base,
+                                data_dir: Optional[Path], metadata: Union[List[Any], Dict[str, Any]]):
+    """
+    Get the expected output values with respect to the output types.
+    This is done to get the md5sum of files when converting unit tests.
+
+    Ex: If given hello.txt and type File, try to get the regex of the file
+    If given "hello" and type String, return string
+    """
+    if output_values is None and output_type.optional:
+        return output_values
+    if isinstance(output_type, WDL.Type.File):
+        # this will only populate md5sum if the files exist recursively in the data directory
+        for path in glob.glob(str(data_dir / "**"), recursive=True):
+            if path.endswith(output_values) and os.path.exists(path):
+                with open(path, "rb") as f:
+                    md5sum = hashlib.md5(f.read()).hexdigest()
+                return {'md5sum': md5sum}
+            # else check if there is extra metadata to read from
+            if get_not_none(metadata, "md5sum") is not None:
+                return {'md5sum': metadata.get("md5sum")}
+            if get_not_none(metadata, "regex") is not None:
+                return {'regex': metadata.get("regex")}
+
+    if isinstance(output_type, WDL.Type.StructInstance):
+        converted_output = dict()
+        for i, (output_key, output_value) in enumerate(output_values.items()):
+            try:
+                output_value_type = output_type.members[output_key]
+            except KeyError:
+                # coercion from map to struct can cause some weird type behavior
+                # since dictionaries past python 3.6 are ordered, find the corresponding type from the current output's position
+                output_value_type = list(output_type.members.values())[i]
```

Comment:
This seems like it's covering up a problem that shouldn't be happening in a way that isn't guaranteed to work. Why would the types have been inserted in the same order as the values? Why would the types not be under the right keys? If we can't reliably do key lookup, why don't we just take lists instead and actually specify the ordering constraint?

Reply:
This is covering up a problem from MiniWDL as it can parse a [map coerced to a struct](https://github.com/openwdl/wdl/blob/20ff54d261e44e903d94149eb1f5c4dd141d26ea/SPEC.md?plain=1#L1425-L1455) but does not properly propagate the keys. This results in the key lookup failing even though the corresponding item types are still correct. This is mainly taking advantage of dictionaries in python always being ordered and miniwdl processing things sequentially.

I'm not sure I can do much about this as the parsing is a MiniWDL issue, and creating my own parser would be quite difficult to deal with just one problem.

Alternatively, I could always use ordering to find the associated type from the member name, but I think it's a little odd compared to doing a lookup when I can.

Reply:
I thought I did create an issue, but it appears I forgot.

Reply:
Added a comment